upload-api/otel.js

```diff
@@ -0,0 +1,123 @@
+import { context as otContext, propagation, trace, SpanStatusCode } from '@opentelemetry/api'
+import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node'
+import {
+  AlwaysOffSampler,
+  AlwaysOnSampler,
+  BatchSpanProcessor,
+  ParentBasedSampler,
+  TraceIdRatioBasedSampler,
+} from '@opentelemetry/sdk-trace-base'
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'
+import { Resource } from '@opentelemetry/resources'
+import {
+  SEMRESATTRS_DEPLOYMENT_ENVIRONMENT,
+  SEMRESATTRS_SERVICE_NAME,
+  SEMRESATTRS_SERVICE_VERSION,
+} from '@opentelemetry/semantic-conventions'
+
+const defaultEndpoint = 'https://telemetry.storacha.network:443'
+
+const resource = new Resource({
+  [SEMRESATTRS_SERVICE_NAME]: 'upload-api',
+  [SEMRESATTRS_DEPLOYMENT_ENVIRONMENT]:
+    process.env.SST_STAGE || process.env.NODE_ENV || 'development',
+  [SEMRESATTRS_SERVICE_VERSION]: process.env.VERSION || process.env.COMMIT,
+})
+
+const sampler = createSampler()
+const exporter = new OTLPTraceExporter({
+  url: `${(process.env.OTEL_EXPORTER_OTLP_ENDPOINT || defaultEndpoint).replace(/\/$/, '')}/v1/traces`,
+  headers: parseHeaders(process.env.OTEL_EXPORTER_OTLP_HEADERS),
+})
+
+const provider = new NodeTracerProvider({ resource, sampler })
+provider.addSpanProcessor(new BatchSpanProcessor(exporter))
+provider.register()
+
+const tracer = trace.getTracer('upload-api')
+
+const headerGetter = {
+  keys: (carrier) => (carrier ? Object.keys(carrier) : []),
+  get: (carrier, key) => {
+    if (!carrier) return []
+    const needle = key.toLowerCase()
+    return Object.entries(carrier)
+      .filter(([k]) => k.toLowerCase() === needle)
+      .flatMap(([, v]) => (Array.isArray(v) ? v : [v]))
+      .filter(Boolean)
+  },
+}
+
+/**
+ * Wrap a Lambda handler with a root span.
+ *
+ * @template {(...args: any[]) => any} Fn
+ * @param {string} name
+ * @param {Fn} handler
+ * @returns {Fn}
+ */
+export const wrapLambdaHandler = (name, handler) =>
+  /** @type {Fn} */ (async (...args) => {
+    const [event] = args
+    // Pick up parent trace context from request headers when present.
+    const carrier = event && typeof event === 'object' ? event.headers : undefined
+    const parentCtx = propagation.extract(otContext.active(), carrier || {}, headerGetter)
+    const span = tracer.startSpan(name, undefined, parentCtx)
+    const ctxWithSpan = trace.setSpan(parentCtx, span)
+
+    try {
+      return await otContext.with(ctxWithSpan, () => handler(...args))
```

Comment:
Apologies I haven't looked into this, but do you know how the handler would pass the parent trace context onwards? I guess it needs to access the request headers?